feat(s3-deployment): implement new signContent option

AMZN-hgoffin opened the pull request:
This adds a new boolean option 'signPayload' to the BucketDeployment construct, so that it can be used with buckets whose IAM policy denies PutObject when x-amz-content-sha256 is not set properly. (See https://docs.aws.amazon.com/AmazonS3/latest/API/bucket-policy-s3-sigv4-conditions.html for example constructions that use this IAM condition key.)

Closes #24711.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (dismissed):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of this pull request does not follow the Conventional Commits format, see https://www.conventionalcommits.org/.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 's3-deployment' instead of 'aws-s3-deployment'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 's3-deployment' instead of 'aws-s3-deployment'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

AMZN-hgoffin commented:
Exemption Request: this feature is a property of the custom resource lambda and not a CloudFormation feature, and so the unit tests provide 100% coverage by 
1. verifying that this config option sets a new environment variable properly
2. verifying that the python handler interprets the environment variable and turns it into an "aws configure set ..." command
3. verifying that the "aws configure set" command is not executed by any other existing handler paths

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9699e3f880f9becf2a67e6e89b1685eff7efbe3d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=27e959c6-84bd-4501-85d6-456a279ce220%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: de8e69e6de6d7643dc21f04f9859003c230d2fc7
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4adfdc06-0de4-42d4-ac05-c69711a7cf20%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cafb3bf6e22f86598cad5ae8692af04338a41a12
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=37ec8b59-4b20-4c6b-894d-abde0452b732%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

AMZN-hgoffin commented:
> > Exemption Request: I do not believe that this commit requires a new integration test because this feature is a property of the custom resource lambda and not a CloudFormation feature. The unit tests provide 100% coverage by
> 
> @AMZN-hgoffin because the feature is a property of the custom resource lambda is exactly why we _do_ need an integration test. We need an integration test that creates a bucket with the specific bucket policy in order to confirm that these changes allow the user to write to the bucket.

Understood, I will put together a complete integration test. I have entirely too much faith in the aws-cli and botocore test suites, I suppose :)

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1da82ebaf2f35028e6b1d3c7d8642f6633622999
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e5787930-9dbe-409c-8e35-3007d64b4541%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 91b295505f15204ef91eb3a1a1be0b07e1a399bc
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4536cfb6-ae8b-4776-a27f-0a0f5175842f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9136983de23f4dfe4fcaba41e770913329d6b6c7
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e944a147-8a96-4ac2-83e8-35811dc62fa9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 115d55ee6b600ca3c98488161616618c32af53f3
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=80d427dc-dafc-46cc-af80-5c1e23310743%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 600b7122bbeae3627ef93a3317c3f639529d856b
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e742194d-cc39-4dff-86df-8d4f223f5f43%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2c28b9830d93f932700bcb9c073d82df78a9045b
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=919b4d09-d8f0-44a9-b392-51aab4d4b0e4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 569f96ebfe649944e3c605233f88a4cd80f5acc1
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2056b4ab-1ed6-4a53-80a3-2eb792106cde%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f4f9dec26f78a4c94a567247a14dd5c61f250743
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ac46b4f9-1fbe-4ca2-ad78-c35e50a1f69a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented on packages/@aws-cdk/aws-s3-deployment/test/integ.bucket-deployment-signcontent.ts:
```diff
@@ -0,0 +1,52 @@
+  testCases: [testCase],
+});
+
+app.synth();
```
This line is not needed when `IntegTest` constructs are used in a test.

AMZN-hgoffin replied:
I copied this from the adjacent tests, can remove it.

TheRealAmazonKendra reviewed (dismissed):
Thank you for your contribution! Please see my comments inline.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

AMZN-hgoffin commented on packages/@aws-cdk/aws-s3-deployment/test/integ.bucket-deployment-signcontent.ts:
```diff
@@ -0,0 +1,52 @@
+      retainOnDelete: false, // default is true, which will block the integration test cleanup
+    });
+    // the test is only valid if the resource policy is already in place
+    deployment.node.addDependency(policyResult.policyDependable);
```
There’s no new policy requirement - all buckets support both signed and unsigned content uploads by default. Setting signContent to true does not necessarily imply that the bucket always requires signed content, just that this deployment construct should use a signed upload.

AMZN-hgoffin replied:
Unfortunately with the flexibility of IAM policies, I think it is impossible to determine from the bucket policy whether signContent needs to be set automatically.

AMZN-hgoffin replied:
Oh, I see. The comment is not saying that the test will fail without the bucket policy. It’s saying that the test can have a false positive deployment success if it doesn’t wait for the policy, because the policy is what restricts unsigned uploads :)

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

AMZN-hgoffin commented:
I think I've addressed the feedback with better comments in the integration test. I rebuilt the change with the new CDK directory structure. Please let me know if there's anything else required to get this functionality landed!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 40c09f8bb5293e589893514dd572e2fc0210d2ef
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=87c84057-53b2-4527-88f1-e06f44ad78a5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

AMZN-hgoffin commented:
Auto-merge was refusing to merge because of workflow files being edited on main since my branch - I have rebased the change to latest main. Diff is identical.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: de823289c9248f78d95d4cd3ddb103604a88c634
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=190a163b-ba65-4d78-b093-94deb11c9987%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d5bf7b3eca6d00d89439e0a4d9ab7a4c566e15f5
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f733f34f-813b-480e-90bb-d1aae781cd38%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 40c09f8bb5293e589893514dd572e2fc0210d2ef
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d798c4c0-6f92-4931-9354-59a44f26e965%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6ce67e90db90efa8363a6e51b5a5f01dca5fa2bf
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2aca249a-05e8-49b2-a2cb-e37be3ed5995%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
